RestCPSubsystemTest.test_promoteExistingCPMember

_master_ (commit 58dc1391368a51933e45519d8f94f4946c445d06)

Failed on Sonar: http://jenkins.hazelcast.com/job/Hazelcast-3.x-sonar/41/testReport/com.hazelcast.internal.ascii/RestCPSubsystemTest/test_promoteExistingCPMember/
Failed on nightly build: http://jenkins.hazelcast.com/job/Hazelcast-3.x-nightly/38/testReport/com.hazelcast.internal.ascii/RestCPSubsystemTest/test_promoteExistingCPMember/

Stacktrace:
```
java.lang.AssertionError: expected:<200> but was:<500>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.internal.ascii.RestCPSubsystemTest.test_promoteExistingCPMember(RestCPSubsystemTest.java:524)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
Hiccups measured while running test 'test_promoteExistingCPMember(com.hazelcast.internal.ascii.RestCPSubsystemTest):'
23:21:55, accumulated pauses: 71 ms, max pause: 4 ms, pauses over 1000 ms: 0
23:22:00, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
23:22:05, accumulated pauses: 516 ms, max pause: 505 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/internal/ascii/RestCPSubsystemTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/ascii/RestCPSubsystemTest.java
@@ -18,6 +18,8 @@ package com.hazelcast.internal.ascii;
 
 import com.hazelcast.config.Config;
 import com.hazelcast.config.JoinConfig;
+import com.hazelcast.config.RestApiConfig;
+import com.hazelcast.config.RestEndpointGroup;
 import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IAtomicLong;
@@ -34,7 +36,6 @@ import com.hazelcast.internal.json.Json;
 import com.hazelcast.internal.json.JsonArray;
 import com.hazelcast.internal.json.JsonObject;
 import com.hazelcast.internal.json.JsonValue;
-import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -71,8 +72,10 @@ public class RestCPSubsystemTest extends HazelcastTestSupport {
 
     @Before
     public void setup() {
-        config.setProperty(GroupProperty.REST_ENABLED.getName(), "true");
-        config.setProperty(GroupProperty.HTTP_HEALTHCHECK_ENABLED.getName(), "true");
+        RestApiConfig restApiConfig = new RestApiConfig()
+                .setEnabled(true)
+                .enableGroups(RestEndpointGroup.CLUSTER_WRITE);
+        config.getNetworkConfig().setRestApiConfig(restApiConfig);
 
         JoinConfig join = config.getNetworkConfig().getJoin();
         join.getMulticastConfig().setEnabled(false);
@@ -484,18 +487,32 @@ public class RestCPSubsystemTest extends HazelcastTestSupport {
     }
 
     @Test
-    public void test_promoteAPMemberToCPMember() throws IOException, ExecutionException, InterruptedException {
-        HazelcastInstance instance1 = Hazelcast.newHazelcastInstance(config);
-        Hazelcast.newHazelcastInstance(config);
-        Hazelcast.newHazelcastInstance(config);
-        HazelcastInstance instance4 = Hazelcast.newHazelcastInstance(config);
+    public void test_promoteAPMemberToCPMember() throws ExecutionException, InterruptedException {
+        final HazelcastInstance instance1 = Hazelcast.newHazelcastInstance(config);
+        final HazelcastInstance instance2 = Hazelcast.newHazelcastInstance(config);
+        final HazelcastInstance instance3 = Hazelcast.newHazelcastInstance(config);
+        final HazelcastInstance instance4 = Hazelcast.newHazelcastInstance(config);
 
-        ConnectionResponse response = new HTTPCommunicator(instance4).promoteCPMember(groupName, groupPassword);
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertNotNull(instance1.getCPSubsystem().getLocalCPMember());
+                assertNotNull(instance2.getCPSubsystem().getLocalCPMember());
+                assertNotNull(instance3.getCPSubsystem().getLocalCPMember());
+            }
+        });
 
-        assertEquals(200, response.responseCode);
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                ConnectionResponse response = new HTTPCommunicator(instance4).promoteCPMember(groupName, groupPassword);
+                assertEquals(200, response.responseCode);
+            }
+        });
 
         Collection<CPMember> cpMembers = instance1.getCPSubsystem().getCPSubsystemManagementService().getCPMembers().get();
         assertEquals(4, cpMembers.size());
+        assertNotNull(instance4.getCPSubsystem().getLocalCPMember());
     }
 
     @Test
@@ -515,9 +532,16 @@ public class RestCPSubsystemTest extends HazelcastTestSupport {
 
     @Test
     public void test_promoteExistingCPMember() throws IOException, ExecutionException, InterruptedException {
-        HazelcastInstance instance1 = Hazelcast.newHazelcastInstance(config);
+        final HazelcastInstance instance1 = Hazelcast.newHazelcastInstance(config);
         Hazelcast.newHazelcastInstance(config);
         Hazelcast.newHazelcastInstance(config);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertNotNull(instance1.getCPSubsystem().getLocalCPMember());
+            }
+        });
 
         ConnectionResponse response = new HTTPCommunicator(instance1).promoteCPMember(groupName, groupPassword);
 

